x/trigger/types/trigger.go

```diff
@@ -0,0 +1,182 @@
+package types
+
+import (
+	fmt "fmt"
+	"strings"
+	time "time"
+
+	proto "github.com/gogo/protobuf/proto"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+type TriggerID = uint64
+
+const (
+	BlockHeightPrefix = "block-height"
+	BlockTimePrefix   = "block-time"
+)
+
+type TriggerEventI interface {
+	proto.Message
+	GetEventPrefix() string
+	Validate() error
+	ValidateContext(ctx sdk.Context) error
+}
+
+var _ TriggerEventI = &TransactionEvent{}
+var _ TriggerEventI = &BlockHeightEvent{}
+var _ TriggerEventI = &BlockTimeEvent{}
+
+// NewTrigger creates a new trigger.
+func NewTrigger(id TriggerID, owner string, event *codectypes.Any, action []*codectypes.Any) Trigger {
+	return Trigger{
+		id,
+		owner,
+		event,
+		action,
+	}
+}
+
+// NewQueuedTrigger creates a new trigger for queueing.
+func NewQueuedTrigger(trigger Trigger, blockTime time.Time, blockHeight uint64) QueuedTrigger {
+	return QueuedTrigger{
+		Time:        blockTime,
+		BlockHeight: blockHeight,
+		Trigger:     trigger,
+	}
+}
```

Comment:
Here too. These constructors would be better placed down with the other funcs/methods for that type.